rmf_traffic/src/rmf_traffic/schedule/Timeline.hpp

```diff
@@ -0,0 +1,517 @@
+            inspector);
+    }
+    else
+    {
+      throw std::runtime_error(
+            "Unexpected Query::Participants mode: "
+            + std::to_string(static_cast<uint16_t>(mode)));
+    }
+  }
+
+  void cull(const Time time)
+  {
+    for (auto& pair : _timelines)
+    {
+      auto& timeline = pair.second;
+
+      // NOTE(MXG): It is not an error that we are using get_timeline_begin() to
+      // find the ending iterator. We want to stop just before the first bucket
+      // that contains the cull time, because we only want to erase times that
+      // come before it.
+      const auto end_it = get_timeline_begin(timeline, &time);
+      if (end_it != timeline.begin())
+        timeline.erase(timeline.begin(), end_it);
+    }
+  }
+
+private:
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_spacetime(
+      const Query::Spacetime& spacetime,
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    const Query::Spacetime::Mode mode = spacetime.get_mode();
+
+    if (Query::Spacetime::Mode::All == mode)
+    {
+      inspect_all_spacetime(participant_filter, inspector);
+    }
+    else if (Query::Spacetime::Mode::Regions == mode)
+    {
+      inspect_spacetime_regions(
+            spacetime.regions(), participant_filter, inspector);
+    }
+    else if (Query::Spacetime::Mode::Timespan == mode)
+    {
+      inspect_spacetime_timespan(
+            spacetime.timespan(), participant_filter, inspector);
+    }
+  }
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_all_spacetime(
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    Checked checked;
+
+    const auto relevant = [](const Route&) -> bool { return true; };
+    for (const auto& entry : _all_bucket)
+    {
+      if (participant_filter.ignore(entry->participant))
+        continue;
+
+      if (!checked[entry->participant].insert(entry->route_id).second)
+        continue;
+
+      inspector.inspect(entry, relevant);
+    }
+  }
+
+  template<typename Inspector, typename ParticipantFilter>
+  void inspect_spacetime_regions(
+      const Query::Spacetime::Regions& regions,
+      const ParticipantFilter& participant_filter,
+      Inspector& inspector) const
+  {
+    Checked checked;
+
+    rmf_traffic::internal::Spacetime spacetime_data;
+    const auto relevant = [&spacetime_data](const Route& r) -> bool {
+      return rmf_traffic::internal::detect_conflicts(
+            r.trajectory(), spacetime_data, nullptr);
+    };
+
+    for (const Region& region : regions)
+    {
+      const std::string& map = region.get_map();
+      const auto map_it = _timelines.find(map);
+      if (map_it == _timelines.end())
+        continue;
+
+      const Entries& timeline = map_it->second;
+      const Time* const lower_time_bound = region.get_lower_time_bound();
+      const Time* const upper_time_bound = region.get_upper_time_bound();
+
```

Comment:
If the region is time bounded, would it be beneficial to check whether it overlaps in time with the given timeline? My understanding is that `rmf_traffic::internal::detect_conflict()` does check for this but only later on in the `inspector() function. Would inserting
```
auto last_it = --timeline.end();
if (lower_time_bound != nullptr && *lower_time_bound > last_it->first)
  continue;
if (upper_time_bound != nullptr && *upper_time_bound < timeline.begin()->first)
  continue;
```
here increase performance in anyway?

Reply:
I might not be thinking about this hard enough, but my instinct tells me that we will effectively get this by the way we iterate from `timeline_begin` to `timeline_end`. If there is no overlap between the time range of the region and the time range of the timeline, then `timeline_begin == timeline_end` and so none of the buckets will be inspected.

Reply:
Ah yes you're right none of the buckets will be inspected. But we will still be looping through each space in the region and calling `inspect_entries()`? Guessing that doesn't cost much?

Reply:
I'm not sure how looping through each space relates to filtering out based on time.

There isn't any bucket system for us to filter based on space, so we're stuck looping through all the regions no matter what; at least until we implement a broadphase partition that accounts for space.

Reply:
Now I understand. Fixed in d7591ee